cmd/skaffold/app/cmd/config/util.go

```diff
@@ -33,22 +33,40 @@ import (
 const defaultConfigDir = ".skaffold"
 const defaultConfigFile = "config"
 
-func resolveKubectlContext() {
+// ResolveKubectlContext sets up the actual kube-context to be used. The order
+// of precedence is:
+// 1) command line
```

Comment:
this is a misunderstanding - there is _no command line flag yet_ for kubectl context. See below, the kubecontext you are using is a global var used to parse the `skaffold config` command's kubeContext flag, so that the `skaffold config` can store config values scoped to a given kubectl context.

Reply:
To clarify, `config.kubecontext` shouldn't be considered as part of this PR, then? Ie, when using commands like `dev` and `run` the only contexts that should be considered are the schema.yaml context and the current kubectl context?

Reply:
right - there are two flows here that are affected ... and as I am digging deeper, this is not that trivial actually. 

1.) "runner" (`skaffold dev/deploy/run`): 

a.) we need some metadata - namely `config.GetLocalCluster` and `GetDefaultRepo` to influence right behavior for artifact cache and image replacement. These both rely on the "right kubectl context". 
b.) We need the "right context" for profile activation. 
c.) the deployer (helm/kubectl) needs to be created with the "right context".
d.) bazel and local docker has some kube context dependent logic. 

Ideally we need a "kubectx.CurrentContext()" to return the "right context" that is not the current kubectx on the machine, but whatever we calculated from config. 

The logic to determine the "right kubectl context" should be based on precedence of `DeployConfig.context` and `kubectl.CurrentContext()` - however there is a bit of a circular dependency now that is problematic because of the profile activations (@dgageot???) - we can theoretically create activate a profile based on the current context that will have a Deploy config pointing to a different context. The way to resolve this would be to implement validation against the skaffold.yaml: the kubecontext under deployConfig should never be different from the profile it is defined under. 

Thus the logic to determine the would be the following: 
- get current kubectl context (from machine) 
- if there exists a profile that is activated by this context - make this the current context for everybody - do validate the deploy configs against mismatching context definition 
- if there is no profile that is activated - check the DeployConfig.context - if it's defined, make that the current context for all the above (a, c, d) clients. 

One other thought I have is that if we add multiple kubecontext support for deployers - we might have to rethink what this global context means - but I think we can do that based on this later - so we can ignore that use case for now. 

**problems with current solution**: 

- does not need the "kubecontext" flag from `skaffold config` - as it is never called. 
- the fact that you are storing the result of the resolution in a flag's value that is disconnected also should be reconsidered as everyone else is calling `context.CurrentContext()` which will result in inconsistencies - I realize that config package is 

**recommendation**: 
- implement the above logic for determining the kubectx in a function named ResolveAndValidateKubeCtx(config) `pkg/skaffold/kubernetes/context/context.go` 
- this _can_ set a state variable in `pkg/skaffold/kubernetes/context`
- `ResolveAndValidateKubeCtx` should be called right after the config has been parsed - maybe even part of the validation there? 
- `context.CurrentContext()`  should always just return the value of the state variable if it's set, if not, it should return the global/machine kubectl context as it does now

2.) `skaffold config` subcommands: have to resolve the right "kubectl context" based on the precedence of `--kube-context` flag (or KUBE_CONTEXT env var -- see `cmd.setFlagsFromEnvVariables`) and `kubectl.CurrentContext()` to retrieve/set/unset the right configuration in the global config. 

**problems with current solution**: 

- does not need the `DeployConfig.context` as `skaffold config` doesn't care about the skaffold.yaml file itself. 

**recommendation**: 
- leave it as is for now, and the rest is going to be taken care of as `context.CurrentContext()`  should always just return the value of the state variable if it's set, if not, it should return the global/machine kubectl context as it does now